util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,461 @@
+	time.Sleep(time.Millisecond * 100)
+	data.normalizedPlanMap.Range(func(key, value interface{}) bool {
+		planDecoded, err := tsr.planBinaryDecoder(value.(string))
+		if err != nil {
+			logutil.BgLogger().Warn("[top-sql] decode plan failed", zap.Error(err))
+			return true
+		}
+		data.normalizedPlanMap.Store(key, planDecoded)
+		return true
+	})
+
+	i := 0
+	batch := make([]*tipb.CollectCPUTimeRequest, len(data.topSQL))
+	for _, value := range data.topSQL {
+		normalizedSQL, sqlOk := data.normalizedSQLMap.Load(string(value.SQLDigest))
+		normalizedPlan, planOk := data.normalizedPlanMap.Load(string(value.PlanDigest))
+		req := &tipb.CollectCPUTimeRequest{
+			TimestampList: value.TimestampList,
+			CpuTimeMsList: value.CPUTimeMsList,
+			SqlDigest:     value.SQLDigest,
+			PlanDigest:    value.PlanDigest,
+		}
+		if sqlOk {
+			req.NormalizedSql = normalizedSQL.(string)
+		}
+		if planOk {
+			req.NormalizedPlan = normalizedPlan.(string)
+		}
+
+		batch[i] = req
+		i++
+	}
+	return batch
+}
+
+// sendToAgentWorker will send a snapshot to the gRPC endpoint every collect interval
+func (tsr *RemoteTopSQLReporter) reportWorker() {
+	defer util.Recover("top-sql", "reportWorker", nil, false)
+
+	for {
+		select {
+		case data := <-tsr.reportDataChan:
+			tsr.Report(data)
+		case <-tsr.quit:
+			return
+		}
+	}
+}
+
+func (tsr *RemoteTopSQLReporter) Report(data reportData) {
+	batch := tsr.prepareReportData(data)
+	addr := variable.TopSQLVariable.AgentAddress.Load()
+	ctx, cancel := context.WithTimeout(context.TODO(), reportTimeout)
+	err := tsr.client.Send(ctx, addr, batch)
+	if err != nil {
+		logutil.BgLogger().Warn("[top-sql] client failed to send data", zap.Error(err))
+	}
+	cancel()
+}
+
+// ReportGRPCClient reports data to grpc servers.
+type ReportGRPCClient struct {
+	addr   string
+	conn   *grpc.ClientConn
+	client tipb.TopSQLAgentClient
+}
+
+// NewReportGRPCClient returns a new ReportGRPCClient
+func NewReportGRPCClient() *ReportGRPCClient {
+	return &ReportGRPCClient{}
+}
+
+// Send implements the ReportClient interface.
+func (r *ReportGRPCClient) Send(ctx context.Context, addr string, batch []*tipb.CollectCPUTimeRequest) error {
+	if addr == "" {
+		return nil
+	}
+	err := r.initialize(addr)
+	if err != nil {
+		return err
+	}
+	// TODO: test timeout behavior
+	stream, err := r.client.CollectCPUTime(ctx)
+	if err != nil {
+		return r.resetClientWhenSendError(err)
+	}
+	if err := r.sendBatch(stream, batch); err != nil {
+		return r.resetClientWhenSendError(err)
+	}
+	return nil
+}
+
+// sendBatch sends a batch of TopSQL records streamingly.
+// TODO: benchmark test with large amount of data (e.g. 5000), tune with grpc.WithWriteBufferSize()
+func (tsr *ReportGRPCClient) sendBatch(stream tipb.TopSQLAgent_CollectCPUTimeClient, batch []*tipb.CollectCPUTimeRequest) error {
+	for _, req := range batch {
+		if err := stream.Send(req); err != nil {
+			return err
```

Comment:
When error happens, should we close the stream?

Reply:
No need to do this, we will close the grpc connection when meet error.